docker/daemon.go

```diff
@@ -119,40 +147,45 @@ func mainDaemon() {
 		serveAPIWait <- nil
 	}()
 
-	registryService := registry.NewService(registryCfg)
-	d, err := daemon.NewDaemon(daemonCfg, eng, registryService)
-	if err != nil {
-		eng.Shutdown()
-		logrus.Fatalf("Error starting daemon: %v", err)
-	}
-
-	if err := d.Install(eng); err != nil {
-		eng.Shutdown()
-		logrus.Fatalf("Error starting daemon: %v", err)
-	}
-
-	logrus.Info("Daemon has completed initialization")
-
-	logrus.WithFields(logrus.Fields{
-		"version":     dockerversion.VERSION,
-		"commit":      dockerversion.GITCOMMIT,
-		"execdriver":  d.ExecutionDriver().Name(),
-		"graphdriver": d.GraphDriver().String(),
-	}).Info("Docker daemon")
+	signal.Trap(func() {
+		api.Close()
+		<-serveAPIWait
+		shutdownDaemon(d)
+		if pfile != nil {
+			pfile.Remove()
+		}
+	})
 
 	// after the daemon is done setting up we can tell the api to start
 	// accepting connections with specified daemon
 	api.AcceptConnections(d)
 
 	// Daemon is fully initialized and handling API traffic
-	// Wait for serve API job to complete
+	// Wait for serve API to complete
 	errAPI := <-serveAPIWait
-	eng.Shutdown()
+	shutdownDaemon(d)
 	if errAPI != nil {
 		logrus.Fatalf("Shutting down due to ServeAPI error: %v", errAPI)
```

Comment:
yes but should be done  here in defer https://github.com/runcom/docker/commit/0bb37c3913cef07e6f4a5ccd54989eeca97093e7#diff-9d18a6929859b5f9739035de6065951dR99 

correct @LK4D4 ? because here everything went well or timeout but function returns and in case execute defer func